File: 2021/golang/day12.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

type caveCrawler struct {
	pos       string
	smallSeen []string
	twice     bool
}

func solveDay12(N map[string][]string, allowTwice bool) int {
	routeCount := 0
	Q := []caveCrawler{{pos: "start", smallSeen: make([]string, 0), twice: false}}
	for len(Q) > 0 {
		crawler := Q[len(Q)-1]
		Q = Q[:len(Q)-1]
		if crawler.pos == "end" {
			routeCount++
		} else {
			for _, n := range N[crawler.pos] {
				sNext := make([]string, len(crawler.smallSeen))
				copy(sNext, crawler.smallSeen)
				if n != "start" && !ContainsString(crawler.smallSeen, n) {
					if IsLower(n) {
						sNext = append(sNext, n)
					}
					Q = append(Q, caveCrawler{pos: n, smallSeen: sNext, twice: crawler.twice})
				} else if ContainsString(crawler.smallSeen, n) && !crawler.twice && n != "start" && n != "end" && allowTwice {
					Q = append(Q, caveCrawler{pos: n, smallSeen: sNext, twice: true})
				}
			}
		}
	}
	return routeCount
}

func Day12() {
	file, _ := os.Open("../input/day12.input")
	defer file.Close()
	N := make(map[string][]string)
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		parts := strings.Split(scanner.Text(), "-")
		a, b := parts[0], parts[1]
		N[a] = append(N[a], b)
		N[b] = append(N[b], a)
	}
	fmt.Println("Answer A:", solveDay12(N, false))
	fmt.Println("Answer B:", solveDay12(N, true))
}

File: 2021/python/day12.py
from collections import defaultdict

N = defaultdict(list)
for line in [i.strip() for i in open("../input/day12.input", encoding="utf-8").readlines()]:
    k, v = line.split("-")
    N[k].append(v)
    N[v].append(k)

def solve(allowTwice: bool):
    routeCount, Q = 0, [("start", set(), False)]
    while Q:
        pos, smallSeen, twice = Q.pop()
        if pos == "end":
            routeCount += 1
        else:
            for n in N[pos]:
                if n not in smallSeen and n != "start":
                    sNext = set(smallSeen)
                    if n.islower():
                        sNext.add(n)
                    Q.append((n, sNext, twice))
                elif n in smallSeen and not twice and allowTwice and n not in ["start", "end"]:
                    Q.append((n, smallSeen, True))
    return routeCount

print("Answer A:", solve(False))
print("Answer B:", solve(True))
